exe/Main.hs

```diff
@@ -1,21 +1,61 @@
 -- Copyright (c) 2019 The DAML Authors. All rights reserved.
 -- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE NamedFieldPuns    #-}
 {-# LANGUAGE OverloadedStrings #-}
-{-# LANGUAGE RecordWildCards   #-}
 module Main(main) where
 
-import           Ide.Arguments (Arguments (..), GhcideArguments (..),
-                                getArguments)
-import           Ide.Main      (defaultMain)
-import           Plugins
+import           Data.Function                ((&))
+import           Development.IDE.Types.Logger (Priority (Debug, Info),
+                                               WithPriority (WithPriority, priority),
+                                               cfilter, cmap,
+                                               withDefaultRecorder)
+import qualified Development.IDE.Types.Logger as Logger
+import           Ide.Arguments                (Arguments (..),
+                                               GhcideArguments (..),
+                                               getArguments)
+import           Ide.Main                     (defaultMain)
+import qualified Ide.Main                     as IdeMain
+import qualified Plugins
+import           Prettyprinter                (Doc, Pretty (pretty))
+import qualified System.Log                   as HsLogger
+
+data Log
+  = LogIdeMain IdeMain.Log
+  | LogPlugins Plugins.Log
+  deriving Show
+
+instance Pretty Log where
+  pretty log = case log of
+    LogIdeMain ideMainLog -> pretty ideMainLog
+    LogPlugins pluginsLog -> pretty pluginsLog
+
+logToPriority :: Log -> Logger.Priority
+logToPriority = \case
+  LogIdeMain log -> IdeMain.logToPriority log
+  LogPlugins log -> Plugins.logToPriority log
+
+logToDocWithPriority :: Log -> WithPriority (Doc a)
+logToDocWithPriority log = WithPriority (logToPriority log) (pretty log)
 
 main :: IO ()
 main = do
-    args <- getArguments "haskell-language-server" (idePlugins False)
+    -- passing mempty for recorder to idePlugins means that any custom cli
```

Comment:
I think the problem I found was to have a logger to pass into `idePlugins` for `getArguments` I need the result of `getArguments` but `getArguments` needs `idePlugins` which needs a logger.

At the time I felt like the most straightforward way is to have getArguments parse any logging arguments first, then create a logger, then parse the remaining arguments but I thought that might be too crazy to change up the argument parsing in the PR, and currently there are no plugins that use the CLI functionality except for the example plugin.

Another way is to have some default logger that doesn't care about command line arguments to pass in for the plugin CLI functionality, but it's kind of weird.

Reply:
I think it would be sane to just pass a dumb stderr logger here? Or comment that we can't pick a logger because we don't know the user's logging preferences yet.

Or, indeed, we could use a dummy log that just accumulates the logs in a `Writer` and then re-log them when we get the real logger!

Reply:
Yeah I think the dumb stderr logger might be OK since for the accumulator version the logs wouldn't appear in real time. Another strategy would be some way of having that plugin cli command receive a logger later using laziness or mutable variable, but I think keeping it simple for now is better.